### PR TITLE
[Broker] Fixed wrong behaviour caused by not cleaning up topic policy service state.

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

The current exception handler does not clean all states, which will cause some wrong behaviour. 

**E.g:**

#### case 1:

https://github.com/apache/pulsar/blob/e20f34bd2a6b4bd7ce72390a593931a855f8e250/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java#L249-L264

When line 259 closes the reader with an exception, it doesn't remove namespace from ``policyCacheInitMap``, so next time this method is called, they will do nothing because ``policyCacheInitMap.putIfAbsent(namespace, false)`` is not empty.
This will cause `getTopicPolicies` throw TopicPoliciesCacheNotInitException.

https://github.com/apache/pulsar/blob/e20f34bd2a6b4bd7ce72390a593931a855f8e250/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java#L193-L205

#### case 2:

When line 384 gets ``AlreadyClosedException``, it doesn't remove namespace from ``policyCacheInitMap``, but the current namespace value in ``policyCacheInitMap`` is true. Therefore, the next time ``prepareInitPoliciesCache `` is called, a new reader is never created. This will cause `getTopicPolicies` get NULL value.

https://github.com/apache/pulsar/blob/e20f34bd2a6b4bd7ce72390a593931a855f8e250/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java#L376-L393

### Modifications

- Clean state when the reader gets an error.
- Don't clean ``ownedBundlesCountPerNamespace`` when reader get ``AlreadyClosedException``.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java [hunk 1]
```diff
@@ -382,10 +382,10 @@ private void readMorePolicies(SystemTopicClient.Reader<PulsarEvent> reader) {
             } else {
                 if (ex instanceof PulsarClientException.AlreadyClosedException) {
                     log.error("Read more topic policies exception, close the read now!", ex);
-                    NamespaceName namespace = reader.getSystemTopic().getTopicName().getNamespaceObject();
-                    ownedBundlesCountPerNamespace.remove(namespace);
-                    readerCaches.remove(namespace);
+                    cleanCacheAndCloseReader(
+                            reader.getSystemTopic().getTopicName().getNamespaceObject(), false);
                 } else {
+                    log.warn("Read more topic polices exception, read again. {}", ex.getMessage());
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java:388]: Stack or message? I' not sure.

**Review by Technoboy- — APPROVED** [review]
LGTM.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java:388]: Yes, it's better to have the stacktrace.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java [hunk 2]
```diff
@@ -373,6 +361,18 @@ private void initPolicesCache(SystemTopicClient.Reader<PulsarEvent> reader, Comp
         });
     }
 
+    private void cleanCacheAndCloseReader(NamespaceName namespace, boolean cleanOwnedBundlesCount) {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java:364]: Do we need to consider the thread-safe issue here?

Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java:364]: Move `policyCacheInitMap.remove(namespace)` to the end of this method to ensure the state.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java (hunk 1 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java]: Fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java [hunk 3]
```diff
@@ -373,6 +361,18 @@ private void initPolicesCache(SystemTopicClient.Reader<PulsarEvent> reader, Comp
...
+        CompletableFuture<SystemTopicClient.Reader<PulsarEvent>> readerFuture = readerCaches.remove(namespace);
+        policyCacheInitMap.remove(namespace);
+        policiesCache.entrySet().removeIf(entry -> entry.getKey().getNamespaceObject() == namespace);
+        if (cleanOwnedBundlesCount) {
+            ownedBundlesCountPerNamespace.remove(namespace);
+        }
+        if (readerFuture != null) {
```
[the remark this answers is not in the document]
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java]: Fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java (hunk 2 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java:364]: @codelipenghui 

Threading issues will be caused by the point in time when reader failures and new reader creations occur at the same time.
We create a new Reader using the "prepareInitPoliciesCache" method, which uses the "ConcurrentHashMap" to ensure that the "policyCacheInitMap" is atomic.
Therefore, we can clean this atomic value last when cleaning to ensure that cleanup and new creation are atomic.
I'm not sure if this is a thread safety issue as you think. Please let me know if I understand wrong.

**Comment by mattisonchao** [comment]
@codelipenghui @Technoboy-  @eolivelli @lhotari  @michaeljmarshall @Jason918 

PTAL, when you have time ~

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java [hunk 4]
```diff
@@ -373,6 +361,18 @@ private void initPolicesCache(SystemTopicClient.Reader<PulsarEvent> reader, Comp
         });
     }
 
+    private void cleanCacheAndCloseReader(NamespaceName namespace, boolean cleanOwnedBundlesCount) {
+        CompletableFuture<SystemTopicClient.Reader<PulsarEvent>> readerFuture = readerCaches.remove(namespace);
+        policiesCache.entrySet().removeIf(entry -> entry.getKey().getNamespaceObject() == namespace);
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java:366]: `equals` instead of `==` ?

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java]: Fixed, PTAL :)

**Comment by lhotari** [comment]
Good work @mattisonchao  
I guess it might be hard to reproduce with tests, but it would be useful to have a test case that reproduces the issue.
Penghui's comment about the thread safety (race condition) is valid. that might require more changes. Tests would help in that case too.

**Comment by mattisonchao** [comment]
> Good work @mattisonchao I guess it might be hard to reproduce with tests, but it would be useful to have a test case that reproduces the issue. Penghui's comment about the thread safety (race condition) is valid. that might require more changes. Tests would help in that case too.

Sure, Very thanks for your suggestions.

I will try to add a multi-thread test to verify race conditions.

**Comment by Technoboy-** [comment]
> 

As @mattisonchao is on holiday this week, how about merging this first and add test later?
@lhotari @Jason918 @codelipenghui

**Comment by codelipenghui** [comment]
> As @mattisonchao is on holiday this week, how about merging this first and adding test later?

Works for me, @JipeiWang could you please help create an issue for tracking the test?

**Review by codelipenghui — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]
LGTM

**Pull request merged**